pkg/mcs/resource_manager/server/grpc_service.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2022 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/pingcap/errors"
+	rmpb "github.com/pingcap/kvproto/pkg/resource_manager"
+	"github.com/tikv/pd/pkg/mcs/registry"
+	"github.com/tikv/pd/server"
+	"google.golang.org/grpc"
+)
+
+var _ rmpb.ResourceManagerServer = (*Service)(nil)
+
+// SetUpRestHandler is a hook to sets up the REST service.
+var SetUpRestHandler = func(srv *Service) (http.Handler, server.APIServiceGroup) {
+	return dummyRestService{}, server.APIServiceGroup{}
+}
+
+type dummyRestService struct{}
+
+func (d dummyRestService) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	w.WriteHeader(http.StatusNotImplemented)
+	w.Write([]byte("not implemented"))
+}
+
+// Service is the gRPC service for resource manager.
+type Service struct {
+	ctx     context.Context
+	manager *Manager
+	// settings
+}
+
+// NewService creates a new resource manager service.
+func NewService(svr *server.Server) registry.RegistrableService {
+	manager := NewManager(svr)
+
+	return &Service{
+		ctx:     svr.Context(),
+		manager: manager,
+	}
+}
+
+// RegisterGRPCService registers the service to gRPC server.
+func (s *Service) RegisterGRPCService(g *grpc.Server) {
+	rmpb.RegisterResourceManagerServer(g, s)
+}
+
+// RegisterRESTHandler registers the service to REST server.
+func (s *Service) RegisterRESTHandler(userDefineHandlers map[string]http.Handler) {
```

Comment:
Do you have any suggestion?

Reply:
I think we can remove `userDefineHandlers` and put a `Server` interface. Because `userDefineHandlers` can be can be a field in the `Server`. After doing this, it seems more simple to move this func to another place

Reply:
`userDefineHandlers` currently needs to be passed to etcd, and also we don't want let `service` to know the `pd.Server` need to deconstruct it later.